array.h

```diff
@@ -2746,6 +2759,37 @@ const_array_ref<T, NewShape> reinterpret_shape(
   return reinterpret_shape(a.cref(), new_shape, offset);
 }
 
+/** Move an array `from` to a new array, reinterpreting the shape of the array
+ * to `new_shape`, with a base pointer offset `offset`. This is only available
+ * for trivial `T`. */
+template <typename NewShape, typename T, typename Shape, typename Alloc>
+array<T, NewShape, Alloc> move_reinterpret_shape(array<T, Shape, Alloc>&& from, const NewShape& new_shape, index_t offset = 0) {
+  // TODO: Use enable_if to implement this check. It is difficult to do due to
+  // the friend declaration.
+  static_assert(std::is_trivial<T>::value, "move_reinterpret_shape is broken for non-trivial types.");
+  assert(new_shape.is_subset_of(from.shape(), offset));
+  array<T, NewShape, Alloc> result;
+  assert(result.alloc_ == from.get_allocator());
```

Comment:
I forget - Allocators are compile-time? So you don't need to `std::move` the allocator?

Reply:
Allocators are weird beasts, and operator== has a particular meaning for them (that they can allocate and free the same pointers). TBH I am not completely sure if allocators should be moved or not, but I'm mirroring the behavior of the existing move constructor here.

In the case of move_reinterpret_shape, we actually can't copy if needed (the allocators are not equal), because we can't do a copy between shapes of different rank. So we need to just error in that case.